Unity/NativeObjectFactory.cs

```diff
@@ -0,0 +1,169 @@
+        }
+
+        public NativeObject GetAnnotationManager()
+        {
+            return new NativeObject(s_GetAnnotationManager(), this, PersistentTypeID.AnnotationManager);
+        }
+
+        public NativeObject GetMonoManager()
+        {
+            return new NativeObject(s_GetMonoManager(), this, PersistentTypeID.MonoManager);
+        }
+
+        public NativeObject Produce(in RuntimeTypeInfo type, int instanceID, ObjectCreationMode creationMode)
+        {
+            // TODO: Support producing abstract types. To do this, the following steps are necessary:
+            //       1. Replace T::VirtualRedirectTransfer with T::Transfer. This can be done by either
+            //          hooking the method via EasyHook, or modifying the virtual function table.
+            //          This works because both methods have compatible signatures.
+            //       2. Create a new Factory method for the type, by locating its constructor function
+            //          and using that to create a new delegate.
+            //       3. Create a new RuntimeTypeInfo based on the original, with the new Factory method.
+            //          It also needs to have the IsAbstract field set to false.
+            //       4. Hook T::GetTypeVirtualInternal to return the appropriate RuntimeTypeInfo.
+            if (type.IsAbstract)
+                return null;
+
+            // TODO: Why does this take two types?
+            IntPtr ptr;
+            if (IsObjectProduceV1)
+            {
+                ptr = s_ProduceV1(ref type.GetPinnableReference(), instanceID, new MemLabelId(), creationMode);
+            } else
+            {
+                ptr = s_ProduceV2(ref type.GetPinnableReference(), ref type.GetPinnableReference(), instanceID, new MemLabelId(), creationMode);
+            }
+            if (ptr.ToInt64() == 0) return null;
+            return new NativeObject(ptr, this, type.PersistentTypeID);
+        }
+
+        public NativeObject GetOrProduce(in RuntimeTypeInfo type) => type.PersistentTypeID switch
+        {
+            PersistentTypeID.SpriteAtlasDatabase => GetSpriteAtlasDatabase(),
+            PersistentTypeID.SceneVisibilityState => GetSceneVisibilityState(),
+            PersistentTypeID.InspectorExpandedState => GetInspectorExpandedState(),
+            PersistentTypeID.AnnotationManager => GetAnnotationManager(),
+            PersistentTypeID.MonoManager => GetMonoManager(),
+            _ => Produce(type, 0, ObjectCreationMode.Default),
+        };
+
+        public void DestroyIfNotSingletonOrPersistent(NativeObject obj, PersistentTypeID persistentTypeID)
+        {
+            if (obj == null) throw new InvalidOperationException("NativeObject is null");
+            if (s_InstanceIDToObject == null) throw new InvalidOperationException("s_InstanceIDToObject is null");
+            if (s_DestroyImmediate == null) throw new InvalidOperationException("s_DestroyImmediate is null");
```

Comment:
```CSharp
if (obj == null)
    throw new ArgumentNullException(nameof(obj));

// There should probably be a custom UnresolvedSymbolException for these errors.
if (s_InstanceIDToObject == null)
    throw new NotSupportedException($"{nameof(s_InstanceIDToObject)} is null.");

if (s_DestroyImmediate == null)
    throw new NotSupportedException($"{nameof(s_DestroyImmediate)} is null.");
```

Reply:
I'm not sure theses are needed any more, I had null reference exceptions the symbol names changed from ScriptingBackendNativeObjectPtrOpaque to MonoObject, which was part of what prompted me to add exceptions to the symbol resolver, and then I forgot to remove all the null checks I added.

Reply:
Yeah, now that `SymbolResolver` will throw by default, those can probably be removed.